src/rgw/rgw_sal_motr.cc

```diff
@@ -1885,6 +1884,121 @@ int MotrObject::delete_obj_aio(const DoutPrefixProvider* dpp, RGWObjState* astat
   return 0;
 }
 
+int MotrCopyObj_CB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  int rc = 0;
+  ldpp_dout(m_dpp, 20) << "Offset = " << bl_ofs << " Length = " << bl_len << dendl;
+  
+  rc = m_dst_writer->process(std::move(bl), bl_ofs);
+  if(rc < 0){
+    ldpp_dout(m_dpp, 20) << "ERROR: writer process failed rc = " << rc << dendl;
+    return rc;
+  }
+
+  ldpp_dout(m_dpp, 20) << "MotrCopyObj_CB handle_data called rc = " << rc << dendl;
+  return rc;
+}
+
+int MotrObject::copy_object_same_zone(RGWObjectCtx& obj_ctx,
+    User* user,
+    req_info* info,
+    const rgw_zone_id& source_zone,
+    rgw::sal::Object* dest_object,
+    rgw::sal::Bucket* dest_bucket,
+    rgw::sal::Bucket* src_bucket,
+    const rgw_placement_rule& dest_placement,
+    ceph::real_time* src_mtime,
+    ceph::real_time* mtime,
+    const ceph::real_time* mod_ptr,
+    const ceph::real_time* unmod_ptr,
+    bool high_precision_time,
+    const char* if_match,
+    const char* if_nomatch,
+    AttrsMod attrs_mod,
+    bool copy_if_newer,
+    Attrs& attrs,
+    RGWObjCategory category,
+    uint64_t olh_epoch,
+    boost::optional<ceph::real_time> delete_at,
+    std::string* version_id,
+    std::string* tag,
+    std::string* etag,
+    void (*progress_cb)(off_t, void *),
+    void* progress_data,
+    const DoutPrefixProvider* dpp,
+    optional_yield y)
+{
+  int rc = 0;
+  ldpp_dout(dpp, 20) << "Src Object Name : " << this->get_key().get_oid() << dendl;
+  ldpp_dout(dpp, 20) << "Dest Object Name : " << dest_object->get_key().get_oid() << dendl;
+  std::unique_ptr<rgw::sal::Object::ReadOp> read_op = this->get_read_op(&obj_ctx);
+
+  std::shared_ptr<rgw::sal::Writer> dst_writer = store->get_atomic_writer(dpp, y,
+          std::unique_ptr<rgw::sal::Object>(dest_object),
+          user->get_id(), obj_ctx,
+          &dest_placement, olh_epoch, *version_id);
+
+  // prepare write op
+  rc = dst_writer->prepare(y);
```

Comment:
Do the read prepare first before creating the write object.

Reply:
Does the sequence matter? or is it just for readability?

Reply:
Done